### PR TITLE
Typed support for SIZE parameter (RFC 1870)

### PR DESCRIPTION
**Pull request by zeenix**
This PR main adds typed support for SIZE parameter but also:

* makes `Parameter` type an enum so future type support for other parameters can be more easily added. ~~Note that such addtions would still be an API-breaks though since the enum is exhaustive (I'm happy to change that as well if desired).~~ Note that the enum is non-exhaustive so that future variant addtions would not cause API breaks.
* Fixes a bunch of clippy warnings.

### CODE REVIEW

File: src/parse/command.rs
Comment (duesee) [inline src/parse/command.rs]: Can we match case-insensitive here?

**Review by duesee — CHANGES_REQUESTED** [review]
Hey, and thanks for the PR! This looks very good. I would only make the suggestion to match case-insensitive on `"SIZE"`. Regarding breaking the API due to exhaustiveness: Good point. But I think we can address that later when we start to care about API breakages. Maybe using `#[non_exhaustive]` is all we need. But there are some tradeoffs we maybe want to discuss.

**Comment by duesee** [comment]
Don't mind the failing CI. This will be fixed in https://github.com/duesee/smtp-codec/pull/5.

File: src/parse/command.rs
Comment (zeenix) [reply src/parse/command.rs]: >  I would only make the suggestion to match case-insensitive on "SIZE". 

Actually that's exactly what I just did before reading your comment here cause it brakes my manual test. :)

> Regarding breaking the API due to exhaustiveness: Good point. But I think we can address that later when we start to care about API breakages

Sure, only thing is we need to remember. :-D IMHO it's best to handle such things before they're actually needed.

> Maybe using #[non_exhaustive] is all we need.

Yeah.

> But there are some tradeoffs we maybe want to discuss.

Given that we've a catch-all case already, I think in this case non_exhaustive makes sense but your call if we want to do that now or later.

Comment (duesee) [reply src/parse/command.rs]: > Sure, only thing is we need to remember. :-D IMHO it's best to handle such things before they're actually needed.

Yeah :-) Actually... I think that it might really be a good idea to make it non_exhaustive right away. So, feel free to do so now!

Comment (zeenix) [reply src/parse/command.rs]: > feel free to do so now!

Done.

**Comment by zeenix** [comment]
@duesee I also added an additional commit now to declare the existing open-ended enums as non-exhaustive. Let me know if you disagree and I'll remove that commit.

**Review by duesee — APPROVED** [review]
All changes are good :-) Thanks!

**Comment by zeenix** [comment]
> All changes are good :-) Thanks!

Awesome. Anything left before it can be merged?

**Comment by duesee** [comment]
🎉

**Pull request merged**
